### DISCUSSION TITLE
Child Plugin ?

### DISCUSSION THREAD

**Discussion by celltower** [gdarko/digital-license-manager#42] [General]
how is it possible to add costum datasabe colums? 
My Problem: I need to add some costum types to the Licens Key but I want to recive updates.
I allready did a fork of the olf Plugin but I want to use the newer one and recive updates..
What would you recommend to do?

**Comment by gdarko** [comment]
Hey @benisdev-py  - Best way is to utilize the dlm_licenses_meta table and save your custom data here.

You have all the available functions for working:

1. Add data
https://github.com/gdarko/digital-license-manager/blob/master/helpers.php#L238

2. Update data
https://github.com/gdarko/digital-license-manager/blob/master/helpers.php#L238

3. Delete Data
https://github.com/gdarko/digital-license-manager/blob/master/helpers.php#L238

5. Get data
https://github.com/gdarko/digital-license-manager/blob/master/helpers.php#L238

Let me know if you have any other questions.

**Reply by celltower** [reply]
Thank you @gdarko

Oh okay, so I can add new Colums like "level1" or "randomnr" as a Column. in the "wpeg_dlm_license_meta" and than I can get the Costum Colums via the (4. Get data part) ? 

And when I use the "dlm_get_license_meta" I can get all of my Costum  Columns back?

**Reply by gdarko** [reply]
Hey @benisdev-py  - I don't advise to add columns to the plugin's tables. Just like it's strongly discouraged to add custom columns on the WordPress native tables like wp_posts, wp_postmeta, etc.

I suggest adding your license related data to the wp_dlm_licenses_meta table using the above functions. This table acts as key:value store associated to specific license id. So you can add unlimited amount of data pieces for specific licnese.

So you will need to adjust your software to write the additional data about specific license to the licenses_meta table.

**Reply by gdarko** [reply]
To migrate the custom columns, just write a simple script that will iterate over the old licenses table and transfer the custom data you have there to digital license manager. You can execute it in any way you want (via browser or command line)

Assuming that you used the Digital License Manager's migration tool and selected "**Preserve Old Ids**"

```
global $wpdb;
$table = $wpdb->prefix . 'lmfwc_licenses';
$licenses = $wpdb->get_results("SELECT * FROM $table");
foreach($licenses as $license) {
    $custom_data_column_value1 = $license->custom_data_column1;
    $custom_data_column_value2 = $license->custom_data_column2;
    dlm_update_license_meta($license->id, 'custom_data_column1', $custom_data_column_value1);
    dlm_update_license_meta($license->id, 'custom_data_column2', $custom_data_column_value2);
}
```

The code is untested but should work with small adoptions.

**Important**: The code assumes that you used "Preserve Old Ids" when migration and your license IDS are the same as your License Manager for WooCommerce database. 

@benisdev-py 

https://docs.codeverve.com/digital-license-manager/migration/migrate-from-license-manager-for-woocommerce/

**Reply by celltower** [reply]
thank you very much for this massiv help!! thats helped me a lot. I now have added like 15 custom data columns.
But I can get and update or edit them easy over the `dlm_update_license_meta($license->id, "custom_column1", "custom_column2", "custom_column3", "custom_column4 "); `

**Reply by gdarko** [reply]
@benisdev-py - 

1. If you don't query the data in SELECT queries, you can also serialize it in a single column. This way you will avoid adding a lot of columns to the database, but performance may not be ideal again, can be just better than multiple meta rows, if you have a LOT of data.

Like:
```
$data = ['key1' => 'value 1', 'key 2' => 'value 2',  'key15' => 'value 15'];
$data = maybe_serialize($data);
dlm_update_license_meta($license->id, 'additional_data', $data);

```

2. The other option is to create a separate table to hold this additional data and keep one additional row for each license

like:



```
wp_dlm_additional_license_data:

- licnese_id
- col_name1
- col_name2
- col_name3
- col_name15
```

Then write your own functions for C/R/U/D operations:

```
global $wpdb;

// Example Read operation
$additional_data = $wpdb->get_row($wpdb->prepare("SELECT * FROM {$wpdb->prefix}dlm_additional_license_data WHERE license_id=%d", $license->id));

// For the other operations use: (I suggest creating simple PHP functions that can be reused for your purpose)
// $wpdb->update()
// $wpdb->insert()
// $wpdb->delete()

```
This way you get very small performance impact and you can join the data with the licenses table easily, also this is very future proof too.

**Reply by gdarko** [reply]
For transfering the data faster, you can define the custom columns in separate `$columns` variable and do a sub-loop:

```
$columns = ['col1', 'col2', 'col3'];
global $wpdb;
$table = $wpdb->prefix . 'lmfwc_licenses';
$licenses = $wpdb->get_results("SELECT * FROM $table");
foreach($licenses as $license) {

    foreach($columns as $column) {
       if(!empty($license->$column)) {
          dlm_update_license_meta($license->id, $column, $license->$column);
       }
    }
}
```

**Reply by gdarko** [reply]
The `dlm_update_license_meta` function will either update existing or create a new meta entry if it doesn't exists by the ID, KEY match.

**Reply by celltower** [reply]
@gdarko Oh Okay, I think I will try Option 2 with the additional Table. I will try all of your very helpful tips and inform you if I got it running :) 
Really Thank you for this Help :=)

**Reply by gdarko** [reply]
You are welcome @benisdev-py

**Reply by celltower** [reply]
Thank you it worked :) 
I bought the PRO Version how long dose it take till I get the ZIP? or the key :)

**Reply by celltower** [reply]
@gdarko Hey, no another question I would like to address the Costum Colums via the api in how far would that be possible? to get them as well as to edit them

**Reply by gdarko** [reply]
@benisdev-py  - You have a action hook here:
https://github.com/gdarko/digital-license-manager/blob/master/templates/admin/licenses/page-edit.php#L220

And you can add your own fields something like this:

```
add_action('dlm_admin_license_form_end', function($license){

   // Load the custom value here from the additional table data.
    global $wpdb;
    $licenseId = $license->getId();
    $additionalData = $wpdb->get_row($wpdb->prepare('SELECT * FROM additional_licenses_data WHERE license_id=%d'));
    $column1value = $additionalData ? $additionalData->column1 : '';
  

    ?>
    <!-- Custom Column 1 -->
    <tr scope="row">
      <th scope="row">
          <label for="column1"><?php esc_html_e('Custom column', 'digital-license-manager');?></label>
      </th>
      <td>
          <input name="column1" id="column1" class="regular-text" type="text" value="<?php echo esc_attr($column1value); ?>">
          <p class="description" id="tagline-description">
              <strong><?php esc_html_e('Optional.', 'digital-license-manager');?></strong>
              <span><?php _e('Your field description text here...');?></span>
          </p>
      </td>
   </tr>
  <?php

});
```

To save the additional fields you can do something like this:

```
add_action('admin_post_dlm_update_license_key', function(){

    check_admin_referer( 'dlm_update_license_key' );

    if ( current_user_can( 'dlm_edit_licenses' ) ) {

        // Save your data here.
        global $wpdb;
        $licenseId   = isset( $_POST['license_id'] ) ? absint( $_POST['license_id'] ) : null;

        $custom1value = isset($_POST['custom1']) ? sanitize_text_field($_POST['custom1']) : '';
        // Run $wpdb->update();
    }
}, 9);
```

**Important**: This code is untested but i think should work, please test and adopt it as per your needs.

**Reply by celltower** [reply]
@gdarko okay nice this is a good option. Can I do a child them of the license add and license edit page?

**Reply by celltower** [reply]
@gdarko hey i know maybe this does not go directly to the plugin but i want to make sure that the plugin works properly i want to change the url and the name of "License" in the MyAccount so the URL and the name in the table of MyAccount how exactly do i do this without breaking anything? 

edit:
I have seen that under "includes/Integrations/WooCommerce/MyAccount.php" you can add costumes Rewrites but when I update the plugin the rewrite is overwritten or?

**Reply by gdarko** [reply]
Hey @benisdev-py  - There is no easy way for now, but i can include a filter for the next release that allows to modify the digital-license endpoint to something else.

If you had the other plugin before, you can simply add redirect for those links to the links of our plugin?

**Reply by celltower** [reply]
@gdarko Okay this would be nice if you can add this filter :) what do you think when the next release is?

**Reply by gdarko** [reply]
Hey @benisdev-py  - Currently, on the insert/create page, we don't have such action hook, however this will be added in our next release.

Hopefully in the next 2-3 weeks.

**Comment by gdarko** [comment]
Done. Thank you, just approved. I manually approve orders at this time due to abuse.On 29.3.2024, at 02:10, Benisdev ***@***.***> wrote:﻿
Thank you it worked :)
I bought the PRO Version how long dose it take till I get the ZIP? or the key :)

**Reply by celltower** [reply]
@gdarko 

To get the custom lcienses colum I do this: this is working for me fine: 

```
function get_customcolum1($license) {
    global $wpdb;
    $licenseId = $license->getId();
    $additionalData = $wpdb->get_row($wpdb->prepare('SELECT * FROM wp_dlm_license_meta_custom WHERE license_id = %d', $licenseId));
    return $additionalData ? $additionalData->customcolum1 : '';
}

add_action('dlm_admin_license_form_end', function($license) use (&$customcolum1) {
    $customcolum1 = get_customcolum1($license);
});

```



To save the additional fields I do this but this is not working I am not sure if the update button can not execute the update functions or what problem there is I can not see any error in Browser or in php error log:
```

add_action('admin_post_dlm_update_license_key', function($license) use (&$customcolum1){
    
    if (!isset($_POST['_wpnonce']) || !wp_verify_nonce($_POST['_wpnonce'], 'dlm_update_license_key')) {
        die('Security check failed');
    }
    

    global $wpdb;

    $licenseId   = isset( $_POST['license_id'] ) ? absint( $_POST['license_id'] ) : null;

    $customcolum1 = isset( $_POST['customcolum1'] ) ? sanitize_text_field($_POST['customcolum1']) : '';

    $result = update_customcolum1($licenseId, $customcolum1);
    
    if ($result) {
        echo '<script>console.log("Database update result: Success");</script>';
    } else {
        echo '<script>console.log("Database update result: Failure");</script>';
    }
}

, 9);


function update_customcolum1($licenseId, $customcolum1) {

    global $wpdb;
    $table_name = $wpdb->prefix . 'dlm_license_meta_custom';

    $data = array(
        'customcolum1' => $customcolum1,
    );

    $where = array(
        'license_id' => $licenseId,
    );

    $result = $wpdb->update( $table_name, $data, $where );

    return $result;
}
```

**Reply by gdarko** [reply]
Hey @benisdev-py  - I think using that console.log statement is wrong and won't output anything.

Please replace the entire `echo()` with `error_log()` and [enable debug logging](https://docs.codeverve.com/digital-license-manager/how-to-enable-debug-log/). This way the logged messages will be available in wp-content/debug.log.

Also, you can print a simple `die('I am here');` statement somewhere before the nonce check or after the nonce check, to see if it reaches that piece of code at all.

**Reply by celltower** [reply]
@gdarko  Okay, i can not see any error in there but I am also not sure if the insert to the DB is korrekt or if I have to add some action at the update button section

**Reply by celltower** [reply]
I think that the "Update" button is not trgierring the function maybe the name is not correct ? 

My Submit button :

`<p class="submit">
                <button name="submit" id="edit__submit" class="button button-primary" value="1" type="submit">
                    <?php esc_html_e( 'Update', 'digital-license-manager' ); ?>
                </button>
            </p>
`

**Reply by gdarko** [reply]
I will test this and get back to you asap.

**Reply by celltower** [reply]
@gdarko did you find any thing how I can edit this custom Paramter in the DB?

**Reply by gdarko** [reply]
Hey @benisdev-py - i tested the hook `admin_post_dlm_update_license_key` and executes correctly on my end. Looks like  you are doing something wrong, if possible please forward the code at info@codeverve.com and i will check.

**Reply by celltower** [reply]
@gdarko Okay I have written your a email :=)
